docs/design/operator.md

```diff
@@ -0,0 +1,407 @@
+  single-namespace mode.
+
+## Development Methods
+
+Use operator framework which is higly used in operator developed and also by
+following Kubernetes best practices for operator development.
+
+## Architecture
+
+```mermaid
+graph TD
+    A[CRD Changes] --> B[Operator]
+    B -->  D[Configure CephFS CSI Driver]
+    B -->  E[Configure RBD CSI Driver]
+    B -->  F[Configure NFS CSI Driver]
+
+```
+
+In this updated diagram:
+
+- **CRD Changes**: Represents changes made to Custom Resource Definitions,
+  which trigger actions in the operator.
+- **Operator**: Listens for changes in CRDs and initiates the installation and
+  configuration of CSI drivers.
+- **Configure CephFS, NFS, RBD**: Actions performed by the operator to install
+  and configure the respective CSI drivers based on the CRD changes.
+
+## CRD's for ceph-csi-operator
+
+### CephCSIOperatorConfig CRD
+
+Manages operator-level configurations and offers a place to overwrite settings
+for CSI drivers. This CRD is a namespace-scoped CRD and a single CR named
+ceph-csi-operator-config can be created by the admin in the namespace
+where operator is running. The configurations are catagorized into 2 different
+parts
+
+- Operator configurations
+- Common configurations across all CSI driver
+
+```yaml
+---
+kind: CephCSIOperatorConfig
+apiVersion: csi.ceph.io/v1alpha1
+metadata:
+  name: csioperatorconfig
+  namespace: operator-namespace
+spec:
+  logLevel: 1
+  driverSpecDefaults:
+    logging:
+      logLevel: 5
+      maxfiles: 5
+      maxLogSize: 10M
+    clusterName: 5c63ad7e-74fe-4724-a511-4ccdc560da56
+    enableMetadata: true
+    grpcTimeout: 100
+    snapshotPolicy: auto-detect
+    generateOMapInfo: true
+    fsGroupPolicy: File
+    encryption:
+      configMapRef:
+        name: encryption-config-map-name
+    plugin:
+      priorityClassName: system-node-critical
+      updateStrategy:
+        type: RollingUpdate
+        rollingUpdate:
+          maxUnavailable: 1
+      labels:
+        app: csi
+      annotations:
+        k8s.v1.cni.cncf.io/networks: macvlan-conf-1
+      affinity:
+        nodeAffinity:
+          requiredDuringSchedulingIgnoredDuringExecution:
+            nodeSelectorTerms:
+            - matchExpressions:
+              - key: storage
+                operator: In
+                values:
+                - node
+      tolerations:
+      - key: storage
+        operator: Exists
+      resources:
+        registrar:
+          limits:
+            cpu: '200'
+            memory: '500'
+          requests:
+            cpu: '100'
+            memory: '250'
+        liveness:
+          limits:
+            cpu: '200'
+            memory: '500'
+          requests:
+            cpu: '100'
+            memory: '250'
+        plugin:
+          limits:
+            cpu: '200'
+            memory: '500'
+          requests:
+            cpu: '100'
+            memory: '250'
+      volumes:
```

Comment:
The volumes need to be configurable? Looking at the current csi templates, I only see the kubelet dir as being configurable in the csi driver volumes, though I didn't look very deeply.

Reply:
Yes we do have this flexibility in Rook as very based on different environments, sometime admin needs additional mounts to the csi driver. https://github.com/rook/rook/blob/8dc88c39b105c79f8f5678f83dd23414411daa40/deploy/examples/operator.yaml#L168-L196